ingester/src/arcmap.rs

```diff
@@ -0,0 +1,331 @@
+//! A map key-value map where values are always wrapped in an [`Arc`], with
+//! helper methods for exactly-once initialisation.
+
+#![allow(dead_code)]
+
+use std::{
+    borrow::Borrow,
+    hash::{BuildHasher, Hash, Hasher},
+    sync::Arc,
+};
+
+use hashbrown::{
+    hash_map::{DefaultHashBuilder, RawEntryMut},
+    HashMap,
+};
+use parking_lot::RwLock;
+
+/// A key-value map where all values are wrapped in [`Arc`]'s and shared across
+/// all readers of a given key.
+///
+/// Each key in an [`ArcMap`] is initialised exactly once, with subsequent
+/// lookups being handed an [`Arc`] handle to the same instance.
+#[derive(Debug)]
+pub(crate) struct ArcMap<K, V, S = DefaultHashBuilder> {
+    map: RwLock<HashMap<K, Arc<V>, S>>,
+    hasher: S,
+}
+
+impl<K, V, S> std::ops::Deref for ArcMap<K, V, S> {
+    type Target = RwLock<HashMap<K, Arc<V>, S>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.map
+    }
+}
+
+impl<K, V> Default for ArcMap<K, V> {
+    fn default() -> Self {
+        Self {
+            map: Default::default(),
+            hasher: Default::default(),
+        }
+    }
+}
+
+impl<K, V, S> ArcMap<K, V, S>
+where
+    K: Hash + Eq,
+    S: BuildHasher,
+{
+    /// Fetch an [`Arc`]-wrapped `V` for `key`, or if this is the first lookup
+    /// for `key`, initialise the value with the provided `init` closure.
+    ///
+    /// # Concurrency
+    ///
+    /// This call is thread-safe - if two calls race, a value will be
+    /// initialised exactly once (one arbitrary caller's `init` closure will be
+    /// executed) and both callers will obtain a handle to the same instance of
+    /// `V`. Both threads will eagerly initialise V and race to "win" storing V
+    /// in the map.
+    ///
+    /// # Performance
+    ///
+    /// This method is biased towards read-heavy workloads, with many readers
+    /// progressing in parallel. If the value for `key` must be initialised, all
+    /// readers are blocked while `init` executes and the resulting `V` is
+    /// memorised.
+    pub(crate) fn get_or_else<Q, F>(&self, key: &Q, init: F) -> Arc<V>
```

Comment:
I would personally find a name like `get_or_insert_with` clearer to understand the intent (as map can be modified and `self` is not `mut`)  though I can see the rationale to use `or_else` to mirror `Option::unwrap_or_else` and related functions

```suggestion
    pub(crate) fn get_or_insert_with<Q, F>(&self, key: &Q, init: F) -> Arc<V>
```

Reply:
As you point out, [this](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) [is](https://doc.rust-lang.org/std/option/enum.Option.html#method.or_else) [an](https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else) [established](https://doc.rust-lang.org/std/result/enum.Result.html#method.unwrap_or_else) [naming](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or_else) [convention](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else) - I think established naming conventions should beat out personal preference as the latter is... personal! I find the consistent `<action>_or_else()` usage in the rust code base a blessing compared to the mish-mash of names in other languages.

Reply:
I guess I was thinking of following the convention of `std::HashMap`: https://doc.rust-lang.org/stable/std/collections/hash_map/enum.Entry.html#method.or_insert_with

Reply:
Option and HashSet have `get_or_insert_with`.

Reply:
This is a better reason! https://github.com/influxdata/influxdb_iox/pull/6065